### ISSUE TITLE
Model saving, loading, transcribing unseen data

### ISSUE THREAD

**Issue by fauxneticien** [oadams/persephone#59]
Making this issue now before I forget:

Would be great to have docs on:
- Saving model (just save `/exp/n/model` folder somewhere else?)
- Loading model (load saved folder?)
- Running loaded model on unseen data (`model.untranscribe` with loaded model?)

**Comment by oadams** [comment]
Yeah, this is key.

**Comment by nikopartanen** [comment]
I work currently in [IKDP-2](https://langdoc.github.io/IKDP-2/) research project, where we are testing different tools with our Komi-Zyrian data. This could be very useful for us. I already did the training part succesfully, and was indeed wondering what's the best way to go onward from that. I'll keep my eye on the repository! Thanks for very interesting and inspirational work!

**Comment by shuttle1987** [comment]
Regarding configuration how much should we aim to include in this PR: https://github.com/oadams/persephone/pull/34/files ?

**Comment by oadams** [comment]
Hi Niko, great to see your interest! The Komi-Zyrian documentation project looks promising. What sort of speech have you collected?

@fauxneticien @nikopartanen I've written a bit of an explanation on this in section 5 [here](https://github.com/oadams/persephone/tree/v0.1.8-dev). Going through it its clear I need to brush up a few things before the docs can be included in the master branch. Let me know how you get on with it!

**Comment by nikopartanen** [comment]
Hi Oliver! Majority of the data is conversations and interviews between native speakers, although usually one speaker is the most dominant. Older recordings tend to have more monologues. There is lots of overlapping speech, and short discoursive elements such as "mm" and "aha" are very common. Everything is rather accurately transcribed on phoneme level using transcription system that matches closely with cyrillic Komi orthography. This can be converted quite easily to some other system that marks phonemes as separate units if that helps (Cyrillic orthography is not easily segmentable on that level). 

All in all we have around 35 hours of transcribed segments. The oldest recordings are from the 50s, but the majority of data is from last ten years. There is quite much variation in recording quality as there are so many different sources. We have tried to include earlier work into our corpus as far as possible, and there are lots of our recordings and archival recordings which are entirely non-annotated, so running even somewhat accurate speech recognition into all that could have many benefits.

The explanation in new section 5 works well. I had to specify my `exp_dir` first as `exp_dir = "exp_kpv"`, then it worked as expected. The only thing I was wondering was whether I can continue the training from the point where I cut it, as I stopped it now after 50 epochs. If I try to run `model.train()` again, it seems to start from epoch 0.

Thanks for great work!

**Comment by oadams** [comment]
> All in all we have around 35 hours of transcribed segments. 

Oh wow, that's quite a lot of data! (by my standards)

> This can be converted quite easily some other system that marks phonemes as separate units if that helps (Cyrillic orthography is not easily segmentable on that level).

This could be good. It would be interesting to compare direct orthographic prediction with the phonemes-as-units model.

Overlapping speech is probably hard to deal with, but fillers and fragments like "mm" and "aha" are not necessarily of a problem. All we can do though is empirically test, I suppose.

> The only thing I was wondering was whether I can continue the training from the point where I cut it, as I stopped it now after 50 epochs. If  I try to run model.train() again, it seems to start from epoch 0.

Cool! What was your error rate at 50 epochs? And was this over the full dataset or a subset? Even though it says "epoch 0", the model _should_ still have the same learnt parameters as before. The best way to test this is just to run it for a single epoch and see if the error rate in that new epoch 0 is similar to where it left off at in the previous epoch 50. Otherwise if the error rate jumps right back up to .90 to 1.0, then we know it's started again.

**Comment by nikopartanen** [comment]
> Oh wow, that's quite a lot of data! (by my standards)

Thanks! We've had several native speakers transcribing for many years in the project, and especially including of archival already transcribed recordings has been important in getting the corpus bigger. To some degree this has been a trade-off with transcription accuracy, and I know that a detailed look into phoneme level reveals many small questions. But of course it is always possible to make a subset and go that through carefully, if that proves to help.

> fillers and fragments like "mm" and "aha" are not necessarily of a problem

Sounds good! I'll test with removing or including these. The segments that contain only these filters can be easily removed. I'm just thinking that the transcription of these doesn't really correspond to normal phoneme sequences in the language, so it may be better if those are not seen during the training. Although who knows.

After 50 epochs validation LER was 0.635851 and PER was 0.743963.

If I do now:

```
model.train()
[stopping manually]
model.transcribe()
model.train()
```

It says it continues from epoch 0 and the scores in log are around 0.96, so it seems to start it from the beginning. I used only a 10 hour subset which I split into training, testing, validation and untranscribed. Now I've done it on my laptop just to get it run but I'll have it later on our laboratory's server.

**Comment by alexis-michaud** [comment]
Hi! If I may jump into the conversation: two fillers are transcribed in the Na texts, as 'mmm...' (nasal filled pause) and 'əəə...' (oral filled pause). Tests show that the model does well at learning patterns of association between these transcriptions and the audio: they are pretty well recognized when transcribing fresh materials. The untranscribed audio on which we wish to run the automatic transcription software has those fillers, so it's much better if the model can deal with them.

So rather than trying to exclude those from the Komi-Zyrian training set, I'd recommend choosing training materials in which those are transcribed (as consistently as possible) and including them in the model.

**Comment by oadams** [comment]
> I'm just thinking that the transcription of these doesn't really correspond to normal phoneme sequences in the language, so it may be better if those are not seen during the training. Although who knows.

They will ultimately be seen at test time though, so it's probably best if the model has had some exposure (I wouldn't bet my life on it though). What I did for Na was create special tokens like "mmm" which were distinct from the token for /m/, effectively training the system to recognize that speaker's fillers. This worked well, although this was a one-speaker system.

> After 50 epochs validation LER was 0.635851 and PER was 0.743963.

Am I right to say that the first number is actually the _training_ LER? By default the validation LER and PER are the same. If I'm wrong though, what is the training LER out of curiosity?

Looking at the code again now, it seems for now you need to pass the `restore_model_path` keyword argument to model.train():

`model.train(restore_model_path="path/to/model_best.ckpt")`

Of course, it would make more sense for the default to work with the existing parameters, unless you explicitly tell it to reset the model. I won't close this issue until that's changed.

**Comment by nikopartanen** [comment]
Thanks for jumping in, @alexis-michaud!

> These units are declared as such, i.e. 'əəə...' is not a sequence of three vowels but one object

I see, this sounds like a very good idea! This is easy to do also with Komi data.

Then @oadams comments:

> Am I right to say that the first number is actually the *training* LER

Yes, exactly, I mixed those up!

> you need to pass the `restore_model_path` keyword argument

This works very well! It overwrites now the log file, and maybe it could be nicer if it would continue writing into it, just to keep track of how it develops. What I get now is:

    Epoch 0. Training LER: 0.664225, validation LER: 0.781792, validation PER: 0.781792

I'll keep it running some time out of curiosity, but I'll set it up soon better with the advices I got here. Thanks!

**Comment by oadams** [comment]
> This works very well! It overwrites now the log file, and maybe it could be nicer if it would continue writing into it, just to keep track of how it develops. 

Good to hear it works. And yes, totally agreed: I also won't close this issue until it instead appends the log.

**Comment by oadams** [comment]
There's a few things to say here.

> Did I understand correctly that | could be used as word boundary?

I haven't experimented using it that way and am quite skeptical of whether it would help at all. It's potentially hurting performance here, but let's find out. Try have the training data a running sequence of phonemes, like: d a a t i a ɕ n ɨ d j a r a n k ɨ v  k o r k e k ɨ v i n n ɨ d.

Another suggestion I'd make is increasing the parameters of your model, since you have so much data. I'd try:

`model = rnn_ctc.Model(exp_dir, na_reader, num_layers=3, hidden_size=400)`

as a starting point. This increases the number of layers in the neural network from 2 to 3, and the number of hidden units per layer from 250 to 400.

> Btw, what does LER exactly mean? PER is phoneme error rate, isn't it?

LER means Label error rate. Anything that is delimited by spaces are "labels". So in this case your labels are phonemes and those word boundary symbols you inserted. The reason the PER is the same is that the system doesn't know what are phonemes and non-phonemes, so it just thinks everything is a phoneme. So for now they are equivalent.

> Would it make sense to train men and women separately? Or to take only better quality recordings from 2000s?

These are hard to answer in a general sense and depend on how well you want the system to generalize. I'd say it's worth experimenting with a few different configurations. I'd be very curious to hear about your findings! 

The main question to ask is whether you will be applying the model to transcribe the speech of a speaker that you already have transcriptions for and can use in training. If so, that is more encouraging of having speaker-specific models. However, if you need to transcribe speakers you haven't seen, you'd likely be better with more general models.

**Comment by nikopartanen** [comment]
Just reporting that I've been now testing this with different configurations. When I removed the word boundary marker and combined some of the primary interjections into individual labels, the best scores changed from:

    Training LER: 0.528446, validation LER: 0.736357, validation PER: 0.736357

To:

    Training LER: 0.553596, validation LER: 0.819382, validation PER: 0.819382

Then I trained it again with the settings you recommended. This took more time (few days), and the scores were:

    Training LER: 0.363763, validation LER: 0.869051, validation PER: 0.869051

I'm happy to run it again with different settings! All the time I've had around 10 hours of material, but I will try it again with the complete corpus. I should also check some of the older recordings which I have there now, although I'm rather sure there are none where transcription would be considerably worse than elsewhere. One could also try to pick the utterances that do not contain overlapping speech, since this is now present quite often.

> The main question to ask is whether you will be applying the model to transcribe the speech of a speaker that you already have transcriptions for and can use in training. 

With my data the model would mainly be applied to entirely new speakers, with the exception of few speakers who are present in most of the recordings (but don't usually speak the most, as they are interviewing / asking questions). So there are lots of different speakers, and usually if there is transcribed data for someone, then we have already finished transcribing all the recordings with that person. Usually there is maybe from 30 minutes to one hour per speaker, although for one speaker we seem to have around 3 hours and 40 minutes (she has often been doing the interviews).

**Comment by oadams** [comment]
Interesting. Given the time it takes to train these models and the poor test scores (86% LER is usually completely wrong), a good strategy now might be to break the corpus down into smaller groupings, such as single-speaker corpora. This makes the task "easier" both in training time and in terms of transcription difficulty. Obviously this isn't your ultimate goal, but for figuring out where the model is having issues this may shed some light.

> Usually there is maybe from 30 minutes to one hour per speaker, although for one speaker we seem to have around 3 hours and 40 minutes (she has often been doing the interviews).

That sounds like a good starting point. Perhaps use only data from this speaker for training, validation and testing. If the model works for her, then we know there's some issue with the remaining data and/or the multi-speaker generalization. If it doesn't work for her, then something more fundamental is going wrong.

Also, how many epochs did those recent training runs go for?
